massa-execution-worker/src/execution.rs

```diff
@@ -1354,4 +1406,95 @@ impl ExecutionState {
+        (
+            self.active_history.read().get_op_exec_status(),
+            self.final_state.read().executed_ops.op_exec_status.clone(),
+        )
+    }
+
+    /// Initialize the hashmap of addresses from the vesting file
+    pub fn init_vesting_registry(
+        config: &ExecutionConfig,
+    ) -> Result<PreHashMap<Address, Vec<VestingRange>>, ExecutionError> {
+        let mut hashmap: PreHashMap<Address, Vec<VestingRange>> = serde_json::from_str(
+            &std::fs::read_to_string(&config.initial_vesting_path).map_err(|err| {
+                ExecutionError::InitVestingError(format!(
+                    "error loading initial vesting file  {}",
+                    err
+                ))
+            })?,
+        )
+        .map_err(|err| {
+            ExecutionError::InitVestingError(format!(
+                "error on deserialize initial vesting file  {}",
+                err
+            ))
+        })?;
+
+        let get_slot_at_timestamp = |config: &ExecutionConfig, timestamp: MassaTime| {
+            match massa_models::timeslots::get_latest_block_slot_at_timestamp(
+                config.thread_count,
+                config.t0,
+                config.genesis_timestamp,
+                timestamp,
+            ) {
+                Ok(opts) => Ok(opts.unwrap()),
+                Err(_) => Err(ExecutionError::InitVestingError(format!(
+                    "can no get the slot at timestamp : {}",
+                    timestamp
+                ))),
+            }
+        };
+
+        for v in hashmap.values_mut() {
+            if v.len().eq(&1) {
+                return Err(ExecutionError::InitVestingError(
+                    "vesting file should has more one element".to_string(),
+                ));
+            } else {
+                *v = v
+                    .windows(2)
+                    .map(|elements| {
+                        let (mut prev, next) = (elements[0], elements[1]);
+
+                        // retrieve the start_slot
+                        if prev.timestamp.eq(&MassaTime::from(0)) {
+                            // first range with timestamp = 0
+                            prev.start_slot = Slot::min();
+                        } else {
+                            prev.start_slot = get_slot_at_timestamp(config, prev.timestamp)?;
+                        }
+
+                        // retrieve the end_slot
+                        let next_range_slot = get_slot_at_timestamp(config, next.timestamp)?;
+                        prev.end_slot = next_range_slot.get_prev_slot(config.thread_count)?;
+
+                        Ok(prev)
+                    })
+                    .collect::<Result<Vec<VestingRange>, ExecutionError>>()?;
+            }
+        }
+
+        Ok(hashmap)
+    }
+
+    /// find a vesting range in the registry, otherwise return None
+    fn find_vesting_range(&self, addr: &Address, current_slot: &Slot) -> Option<&VestingRange> {
```

Comment:
this function is duplicated from context.rs ?

Reply:
yes, because we want to return the VestingError on `execute_roll_buy_op()` without need to acquire the context. 

But i can refactor this to put them after

Reply:
Will be fix in #3699

Reply:
ok but can you add it explicitly to that  issue ? it does not seem to mention this point

Reply:
your comment is linked in the issue (First of "See also")
![image](https://user-images.githubusercontent.com/9607217/226604102-2db99c77-fc7b-4e3a-afd7-55dcfc5678b4.png)

Reply:
sorry missed it :)